### ISSUE TITLE
Advice on how to run the training example in Android

### ISSUE THREAD

**Issue by YuanTingHsieh** [pytorch/executorch#10593]
Hello Team,

We have followed https://pytorch.org/executorch/main/using-executorch-android.html#building-from-source to build the "aar" file.

We can run the inference example on Android.

We are wondering how to run the training example on Android.

Are there some flags / some config we need to add to the building procedure (https://github.com/pytorch/executorch/blob/main/scripts/build_android_library.sh)?

Thanks!

**Comment by tarun292** [comment]
@JacobSzwejbka any thoughts on this?

**Comment by kirklandsign** [comment]
Hi @JacobSzwejbka do we know the cmake config if on linux, not necessarily android?

basically need to copy that to android

**Comment by JacobSzwejbka** [comment]
> Hi [@JacobSzwejbka](https://github.com/JacobSzwejbka) do we know the cmake config if on linux, not necessarily android?
> 
> basically need to copy that to android

https://github.com/pytorch/executorch/blob/main/extension/training/CMakeLists.txt do you mean this?

**Comment by nvkevlu** [comment]
Hi @JacobSzwejbka I have been trying to build and troubleshoot but having trouble getting ExecuTorch built for Android with training.

I have been using the main branch of the executorch code on an Ubuntu 24.04.2 machine to install the dependencies and build ExecuTorch but after much back and forth with versions, I am still stuck. Currently, I cannot even get ExecuTorch to build without Android, and running the following command to build and debug:

`cmake --build cmake-out -j1 --target install --config Release -- VERBOSE=1`

I get the following output with an error for ATen/Tensor.h: No such file or directory:

```
/home/ubuntudev4android/workspace/appforexecutorchtraining/executorch/extension/training/pybindings/_training_lib.cpp:13:10: fatal error: ATen/Tensor.h: No such file or directory
  13 | #include <ATen/Tensor.h>
   |     ^~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [extension/training/CMakeFiles/_training_lib.dir/build.make:79: extension/training/CMakeFiles/_training_lib.dir/pybindings/_training_lib.cpp.o] Error 1
gmake[2]: Leaving directory '/home/ubuntudev4android/workspace/appforexecutorchtraining/executorch/cmake-out'
gmake[1]: *** [CMakeFiles/Makefile2:1268: extension/training/CMakeFiles/_training_lib.dir/all] Error 2
gmake[1]: Leaving directory '/home/ubuntudev4android/workspace/appforexecutorchtraining/executorch/cmake-out'
gmake: *** [Makefile:136: all] Error 2
```

I am using python 3.10 now with 

aiohappyeyeballs       2.6.1
aiohttp                3.11.18
aiosignal              1.3.2
antlr4-python3-runtime 4.9.3
astunparse             1.6.3
async-timeout          5.0.1
attrs                  25.3.0
blobfile               3.0.0
certifi                2025.4.26
charset-normalizer     3.4.2
cmake                  3.31.6
datasets               3.6.0
dill                   0.3.8
exceptiongroup         1.3.0
execnet                2.1.1
executorch             0.7.0a0+d7201ab
expecttest             0.3.0
filelock               3.18.0
flatbuffers            25.2.10
frozenlist             1.6.0
fsspec                 2025.3.0
hf_transfer            0.1.9
huggingface-hub        0.31.2
hypothesis             6.131.16
idna                   3.10
iniconfig              2.1.0
Jinja2                 3.1.6
kagglehub              0.3.12
lark                   1.2.2
lintrunner             0.12.7
lxml                   5.4.0
MarkupSafe             3.0.2
mpmath                 1.3.0
multidict              6.4.3
multiprocess           0.70.16
networkx               3.4.2
ninja                  1.11.1.4
numpy                  2.2.5
omegaconf              2.3.0
optree                 0.15.0
packaging              25.0
pandas                 2.2.3
parameterized          0.9.0
pillow                 11.2.1
pip                    25.1.1
pluggy                 1.5.0
propcache              0.3.1
psutil                 7.0.0
pyarrow                20.0.0
pybind11               2.13.6
pybind11_global        2.13.6
pycryptodomex          3.22.0
pytest                 8.3.5
pytest-rerunfailures   15.1
pytest-xdist           3.6.1
python-dateutil        2.9.0.post0
pytorch-tokenizers     0.1.0
pytz                   2025.2
PyYAML                 6.0.2
regex                  2024.11.6
requests               2.32.3
ruamel.yaml            0.18.10
ruamel.yaml.clib       0.2.12
safetensors            0.5.3
sentencepiece          0.2.0
setuptools             65.5.0
six                    1.17.0
sortedcontainers       2.4.0
sympy                  1.14.0
tabulate               0.9.0
tiktoken               0.9.0
timm                   1.0.7
tokenizers             0.21.1
tomli                  2.2.1
torch                  2.8.0.dev20250325+cpu  /home/ubuntudev4android/workspace/appforexecutorchtraining/venv_py310/lib/python3.10/site-packages
torchao                0.12.0+gitc9b9adc7
torchaudio             2.6.0.dev20250325+cpu
torchdata              0.11.0
torchsr                1.0.4
torchtune              0.6.1
torchvision            0.22.0.dev20250325+cpu
tqdm                   4.67.1
transformers           4.47.1
types-dataclasses      0.6.6
typing_extensions      4.13.2
tzdata                 2025.2
urllib3                2.4.0
wheel                  0.45.1
xxhash                 3.5.0
yarl                   1.20.0
zstd                   1.5.7.0

Do you know if there is something I need to do to resolve the issue with ATen/Tensor.h, or if there a particular version of ExecuTorch or other dependencies I should be trying to build with that might give me a better chance of success?

Thanks!

**Comment by nvkevlu** [comment]
I was able to use 0.6.0 to build with training to run the training example on Ubuntu, but getting the Android build to succeed is continuing to be an issue.

pybind11 seems to be a source of the issues and I'm not sure if this is supposed to be available on Android, but when configuring cmake in order to build ExecuTorch for Android, I run:

cmake -B cmake-out-android-arm64-v8a \
    -DCMAKE_INSTALL_PREFIX=cmake-out-android-arm64-v8a \
    -DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK/build/cmake/android.toolchain.cmake \
    -DANDROID_ABI=arm64-v8a \
    -DANDROID_PLATFORM=android-26 \
    -DCMAKE_BUILD_TYPE=Release \
    -DANDROID_STL=c++_shared \
    -DBUILD_TESTING=OFF \
    -DEXECUTORCH_BUILD_TRAINING=ON \
    -DEXECUTORCH_BUILD_XNNPACK=ON \
    -DEXECUTORCH_XNNPACK_SHARED_WORKSPACE=ON \
    -DEXECUTORCH_BUILD_EXTENSION_MODULE=ON \
    -DEXECUTORCH_BUILD_EXTENSION_TENSOR=ON \
    -DEXECUTORCH_BUILD_EXTENSION_FLAT_TENSOR=ON \
    -DEXECUTORCH_BUILD_EXTENSION_DATA_LOADER=ON \
    -DEXECUTORCH_BUILD_EXTENSION_TRAINING=ON \
    -DEXECUTORCH_BUILD_ANDROID_JNI=ON \
    -DEXECUTORCH_ENABLE_LOGGING=ON \
    -DEXECUTORCH_LOG_LEVEL=Info \
    -DEXECUTORCH_BUILD_KERNELS_OPTIMIZED=ON \
    -DEXECUTORCH_BUILD_KERNELS_QUANTIZED=ON \
    -DEXECUTORCH_BUILD_KERNELS_CUSTOM=ON \
    -DCMAKE_PREFIX_PATH=/home/ubudev4android/workspace/executorch/libtorch \
    -DPYTHON_EXECUTABLE=/home/ubudev4android/workspace/executorch/venv_py310/bin/python \
    -Dpybind11_DIR="/home/ubudev4android/workspace/executorch/venv_py310/lib/python3.10/site-packages/pybind11/share/cmake/pybind11" \
    -DEXECUTORCH_BUILD_DEVTOOLS=ON


but I am now blocked by this error:

CMake Error at /usr/share/cmake-3.28/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Python3 (missing: Python3_INCLUDE_DIRS Python3_LIBRARIES
  Development Development.Module Development.Embed) (found version "3.10.17")
Call Stack (most recent call first):
  /usr/share/cmake-3.28/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.28/Modules/FindPython/Support.cmake:3862 (find_package_handle_standard_args)
  /usr/share/cmake-3.28/Modules/FindPython3.cmake:545 (include)
  extension/training/CMakeLists.txt:61 (find_package)

By the way, for the Ubuntu build I needed to add these lines in extension/training/CMakeLists.txt:

set(Python3_FIND_VIRTUALENV FIRST)
find_package(Python3 COMPONENTS Interpreter Development REQUIRED)
find_package_torch()
find_package(pybind11 REQUIRED)

before:

pybind11_add_module(_training_lib SHARED ${CMAKE_CURRENT_SOURCE_DIR}/pybindings/_training_lib.cpp)

and then I set the flag for cmake -Dpybind11_DIR= to the directory with my pybind11 package for my venv.

**Comment by JacobSzwejbka** [comment]
looking

**Comment by JacobSzwejbka** [comment]
Its strange to me that we are trying to build the python layer here at all lol. 

This should probably be gated https://github.com/pytorch/executorch/blob/main/extension/training/CMakeLists.txt#L44 behind EXECUTORCH_BUILD_PYBIND

**Comment by JacobSzwejbka** [comment]
https://fb.workplace.com/chat/t/9531770920265957 try this @YuanTingHsieh

**Comment by YuanTingHsieh** [comment]
@JacobSzwejbka @kirklandsign Could you share IF android ExecuTorch training is working on the android phones right now?
If not, what is missing?

As far as I know this flag is not helping @nvkevlu 

Thank you very much!
